File: frontend/src/types/telemetry.ts
export type CarGPS = { latitude: number; longitude: number };
export type CarIMU = {
  accelerationX: number; accelerationY: number; accelerationZ: number;
  spinX: number; spinY: number; spinZ: number;
  scale_dps: 250 | 500 | 1000 | 2000;
};
export type CarDrive = { pwm?: number; speed_est_mps?: number };

export type ControlsDerived = {
  steering_deg: number;                       // direita + / esquerda -
  steering_side: 'left'|'right'|'straight';
  speed_cmd_byte: number;                     // 0..255
  speed_cmd_pct: number;                      // 0..1
  speed_cmd_mps: number;                      // m/s
  movement_direction_text: 'front'|'back';
};

export type CentricControls = {
  curve_direction: number;                    // 0..360
  speed: number;                              // 0..255
  movement_direction: 0|1;                    // 0=back, 1=front
  derived?: ControlsDerived;
};

export type TelemetryProcessed = {
  ts: number;                                 // epoch ms
  ts_iso: string;                             // "YYYY-MM-DDTHH:MM:SS.mmmZ"
  ts_local: string;                           // "YYYY-MM-DD HH:MM:SS.mmm±HH:MM"
  src?: string;
  car: { gps?: CarGPS; imu?: CarIMU; drive?: CarDrive };
  centric: { controls: CentricControls };
};

export type TelemetryRaw = {
  received_at: number;
  src?: string;
  raw: {
    car: { gps?: CarGPS; imu?: CarIMU; drive?: CarDrive };
    centric: { controls: Omit<CentricControls, "derived"> };
    src?: string;
  };
};

File: backend/app/schemas/telemetry.py

"""Esquemas Pydantic para validação/serialização de telemetria (versão única)."""
from pydantic import BaseModel, Field, conint, confloat
from typing import Optional

# ---------- Blocos "car" ----------
class CarGPS(BaseModel):
    latitude: confloat(strict=True) = Field(..., description="Latitude WGS-84")
    longitude: confloat(strict=True) = Field(..., description="Longitude WGS-84")

class CarIMU(BaseModel):
    accelerationX: conint(ge=-128, le=127) | int
    accelerationY: conint(ge=-128, le=127) | int
    accelerationZ: conint(ge=-128, le=127) | int
    spinX: conint(ge=-128, le=127) | int
    spinY: conint(ge=-128, le=127) | int
    spinZ: conint(ge=-128, le=127) | int
    scale_dps: conint(strict=True) = Field(..., description="Faixa do giroscópio (°/s): 250, 500, 1000, 2000")

class CarDrive(BaseModel):
    pwm: Optional[conint(ge=0, le=255)] = None
    speed_est_mps: Optional[confloat()] = None  # m/s

class CarBlock(BaseModel):
    gps: Optional[CarGPS] = None
    imu: Optional[CarIMU] = None
    drive: Optional[CarDrive] = None

# ---------- Derivados ----------
class ControlsDerived(BaseModel):
    steering_deg: float
    steering_side: str
    speed_cmd_byte: int
    speed_cmd_pct: float
    speed_cmd_mps: float
    movement_direction_text: str

# ---------- Blocos "centric" ----------
class CentricControls(BaseModel):
    curve_direction: conint(ge=0, le=360)
    speed: conint(ge=0, le=255)
    movement_direction: conint(ge=0, le=1)  # 1=front, 0=back
    derived: Optional[ControlsDerived] = None  # <-- IMPORTANTE para aparecer na resposta

class CentricBlock(BaseModel):
    controls: CentricControls

# ---------- Payload bruto ----------
class TelemetryIn(BaseModel):
    car: CarBlock
    centric: CentricBlock
    src: Optional[str] = Field(default="central")

# ---------- Registro processado ----------
class TelemetryOut(BaseModel):
    ts: int
    ts_iso: str
    ts_local: str
    src: Optional[str]

    car: CarBlock
    centric: CentricBlock